cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/common/DataprocUtils.java

```diff
@@ -292,6 +302,58 @@ private static String getMetadata(String resource) throws IOException {
     }
   }
 
+  /**
+   * Recursively deletes all the contents of the directory and the directory itself.
+   */
+  public static void deleteDirectoryContents(@Nullable File file) {
+    if (file == null) {
+      return;
+    }
+
+    if (file.isDirectory()) {
+      File[] entries = file.listFiles();
+      if (entries != null) {
+        for (File entry : entries) {
+          deleteDirectoryContents(entry);
+        }
+      }
+    }
+    if (!file.delete()) {
+      LOG.warn("Failed to delete file {}.", file);
+    }
+  }
+
+  /**
+   * Recursively deletes all the contents of the directory and the directory itself with retries.
+   */
+  public static synchronized void deleteDirectoryWithRetries(@Nullable File file, String errorMessageOnFailure)  {
```

Comment:
How come we want to retry? Seems a bit odd that deleting a file could fail once, but then succeed soon after. Are permissions or something like that changing behind the scenes?

Reply:
The retry is just to guard against unlikely scenario in which file cannot be deleted (e.g., due to being accessed/locked). Since it is crucial to invalidate cache during initialize, if delete fails, we want to retry before failing the whole provisioner.